### DISCUSSION TITLE
Connectivity issues between strongswan host and Cisco ASA device

### DISCUSSION THREAD

**Discussion by ivanjimenezmedaxion** [strongswan/strongswan#738] [Q&A]
Hi strongSwan community,

We are having difficulty determining why there is sudden instability in our connection to a client's Cisco ASA device. Attached below please find our configuration and the remote Cisco ASA configuration. We have been experiencing intermittent connectivity issues, where our side cannot send traffic to remote subnets.

The security associations from the strongswan side will be online and routable. The security associations from the Cisco side will not be online, however initiating traffic from the Cisco side will bring the security associations up and allow the traffic to flow. 

Cisco client output showing the SAs are not online:
...
AColo-VPN-ASA-01# sh crypto ipsec sa peer [REDACTED] | i ident
      local ident (addr/mask/prot/port): (170.150.78.10/255.255.255.255/0/0)
      remote ident (addr/mask/prot/port): (10.65.74.124/255.255.255.252/0/0)
      local ident (addr/mask/prot/port): (170.150.98.20/255.255.255.255/0/0) // edited by dave
      remote ident (addr/mask/prot/port): (10.65.74.124/255.255.255.252/0/0) // edited by dave
...

In the following output of running "ipsec statusall", the working SAs are 170.150.78.10 and 170.150.98.20 and the remote hosts that cannot be reached are 10.154.26.74/32 and 170.150.226.30/32:
...
Security Associations (2 up, 0 connecting):
         hca[13]: ESTABLISHED 3 hours ago, 10.1.33.12[REDACTED]...199.91.34.69[REDACTED]
         hca[13]: IKEv2 SPIs: 969e61e66ac621c1_i* c09be66dfdedafae_r, pre-shared key reauthentication in 20 hours
         hca[13]: IKE proposal: AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048
       hca11{124}:  INSTALLED, TUNNEL, reqid 12, ESP SPIs: c99c22f1_i 6d667ffc_o
       hca11{124}:  AES_CBC_256/HMAC_SHA2_256_128, 36478656 bytes_i (100802 pkts, 0s ago), 6008579 bytes_o (78463 pkts, 0s ago), rekeying in 4 hours
       hca11{124}:   10.65.74.124/30 === 170.150.78.10/32
       hca13{125}:  INSTALLED, TUNNEL, reqid 14, ESP SPIs: c3c63996_i 62da93a5_o
       hca13{125}:  AES_CBC_256/HMAC_SHA2_256_128, 60890 bytes_i (1052 pkts, 54s ago), 8610990 bytes_o (6346 pkts, 54s ago), rekeying in 7 hours
       hca13{125}:   10.65.74.124/30 === 170.150.98.20/32
         hca[12]: ESTABLISHED 3 hours ago, 10.1.33.12[REDACTED]...199.91.34.69[REDACTED]
         hca[12]: IKEv2 SPIs: ef2c54880903f544_i* 2663df0042097373_r, pre-shared key reauthentication in 20 hours
         hca[12]: IKE proposal: AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_2048
        hca3{119}:  INSTALLED, TUNNEL, reqid 4, ESP SPIs: c1b46585_i 8c877724_o
        hca3{119}:  AES_CBC_256/HMAC_SHA2_256_128, 80080 bytes_i (1339 pkts, 12711s ago), 7714021 bytes_o (8210 pkts, 0s ago), rekeying in 3 hours
        hca3{119}:   10.65.74.124/30 === 10.154.26.74/32
       hca11{120}:  INSTALLED, TUNNEL, reqid 12, ESP SPIs: c6aa4766_i 5326765b_o
       hca11{120}:  AES_CBC_256/HMAC_SHA2_256_128, 2192048 bytes_i (8420 pkts, 3s ago), 469293 bytes_o (7158 pkts, 1026s ago), rekeying in 3 hours
       hca11{120}:   10.65.74.124/30 === 170.150.78.10/32
       hca13{121}:  INSTALLED, TUNNEL, reqid 14, ESP SPIs: c77a3d20_i 1e1750b2_o
       hca13{121}:  AES_CBC_256/HMAC_SHA2_256_128, 150288 bytes_i (2733 pkts, 54s ago), 23874405 bytes_o (25391 pkts, 60s ago), rekeying in 3 hours
       hca13{121}:   10.65.74.124/30 === 170.150.98.20/32
       hca10{122}:  INSTALLED, TUNNEL, reqid 11, ESP SPIs: c39484d2_i 4b769574_o
       hca10{122}:  AES_CBC_256/HMAC_SHA2_256_128, 17759 bytes_i (299 pkts, 13305s ago), 1705480 bytes_o (1330 pkts, 517s ago), rekeying in 4 hours
       hca10{122}:   10.65.74.124/30 === 170.150.226.30/32
...

When we see this condition, more often than not there will be two primary security associations online detected in the "ipsec statusall" output ( see hca[13] and hca[12] ). What does it mean when we see this? Can we prevent this from occurring by modifying any settings on the strongswan side?

Additionally, we would like to know: 
- Would disabling the default NAT-T capability in strongswan be recommended in this scenario? 
- Why would an SA on the strongswan side show as online, but the same SA shows as offline on the Cisco side? 
- Why would initiating traffic from the Cisco side correctly bring up the SA?

Thank you all for any recommendations, ideas or insights.

Regards,

[strongswan_client_config.txt](https://github.com/strongswan/strongswan/files/7498835/strongswan_client_config.txt)
[cisco_client_config.txt](https://github.com/strongswan/strongswan/files/7498838/cisco_client_config.txt)
[strongswan_ip_address.txt](https://github.com/strongswan/strongswan/files/7498862/strongswan_ip_address.txt)
[strongswan_ip_route_show_table_all.txt](https://github.com/strongswan/strongswan/files/7498863/strongswan_ip_route_show_table_all.txt)
[strongswan_ip6tables.txt](https://github.com/strongswan/strongswan/files/7498864/strongswan_ip6tables.txt)
[strongswan_iptables.txt](https://github.com/strongswan/strongswan/files/7498865/strongswan_iptables.txt)

**Comment by tobiasbrunner** [comment]
> What does it mean when we see this? Can we prevent this from occurring by modifying any settings on the strongswan side?

Probably some kind of collision (e.g. both peers creating an SA concurrently initially or maybe due to reauthentication). Not really a problem usually (but don't know if it might be for Cisco). If both ends use traffic based triggers you can't really avoid this from happening, though.

If you don't need it, I'd recommend you disable reauthentication (`reauth=no`), or at least enable make-before-break reauthentication (`charon.make_before_break` in strongswan.conf) as break-before-make reauth will collide with trap policies (`auto=route`) and might cause problems for the responder (the other approach might too, as new SAs are still created from scratch, but overlapping). Without reauthentication, proper rekeying is used, which should also avoid any traffic interruption, see [ExpiryRekey](https://wiki.strongswan.org/projects/strongswan/wiki/ExpiryRekey) for details.

Also note that the combination of `auto=route` and `dpdaction=restart` could cause duplicate CHILD_SAs if the SA goes down and gets restarted while matching traffic concurrently triggers the same CHILD_SA. So `dpdaction=clear` is a better choice (there might be a duplicate anyway if the other peer also uses traffic based triggers - depends on the traffic).

Also note that `dpdtimeout` has no effect on IKEv2 connections ([retransmission timeouts](https://wiki.strongswan.org/projects/strongswan/wiki/Retransmission) are used instead).

> Would disabling the default NAT-T capability in strongswan be recommended in this scenario?

Can't be disabled. Also don't see why it should be relevant.

> Why would an SA on the strongswan side show as online, but the same SA shows as offline on the Cisco side?

No idea. Maybe a log would show some details. But you may have to ask Cisco about that because IKEv2 has proper delete exchanges, so it's strange that a CHILD_SA would be only removed on one side while the IKE_SA is still established.

> Why would initiating traffic from the Cisco side correctly bring up the SA?

Apparently creates the SA it has lost.
